### PR TITLE
Table refresh with ntie

### LINKED ISSUE #4: Have events in Zino web frontend update automatically from backend
Depends on Uninett/zinolib#23

### PR DESCRIPTION
**Pull request by podliashanyk**
Closes #4 

Depends on https://github.com/Uninett/zinolib/pull/53, https://github.com/Uninett/zinolib/pull/58

### Implemented changes:
- Replaced table refresh by polling `get_events()` endpoint with a much more efficient NTIE with `UpdateHandler`. See https://github.com/Uninett/zinolib/pull/53.
- Updated version requirement for the zinolib dependency.
- Refactor: 
    - Renaming - replaced all notion of "poll" in the table refresh logic and config with "refresh". So no more confusing table refresh with "manually poll router or interface"-feature.
    - Extensive cleanup of the old table refresh implementation, including decoupling table refresh from the bulk-event-update response.
- Added HTMX request synchronization via `hx-sync`. This is necessary when dealing with frequent requests to `UpdateHandler` and meanwhile having users potentially updating some events themselves, or selecting/expanding table rows. Without the synchronization, swaps from the "competing" requests were not complete/pending indefinitely, and heap of error alerts showed up in the view. 
- Mitigated #77, the fix is not yet complete though. See https://github.com/Uninett/Howitz/issues/77#issuecomment-2135114070

### Not implemented/further work:
- Dropped the idea of fine-tuning hx-indicator to NTIE table refresh.
With the initial table refresh all table rows during refresh showed spinner indicator at the end of the row like so:
    Screenshot 2024-05-28 at 13 41 40
The thought for NTIE was to only show spinner on the rows that are about to be updated with NTIE (since NTIE updates are more targeted than a refresh of the whole table). Yet I decided to drop this idea and preserve current spinner implementation for following reasons:
    - NTIE refresh might append or remove table rows, so spinners on all rows quite appropriately indicates that **table** is about to change
    - fine-tuning spinners for different NTIE cases is not a one-liner
- Event ordering becomes off with table refresh via NTIE, see https://github.com/Uninett/Howitz/discussions/96
- Rows that are not updated for a while become stale, see https://github.com/Uninett/Howitz/discussions/97

### Attention needed:
- Are any docs updates needed?
- Thoughts on the default refresh interval being reduced to `5`?
- Should we have a minimum refresh interval value?
- Manual tests are welcome: adding, removing, modifying events on Zino server and then checking whether Howitz table is properly updated

### CODE REVIEW

File: src/howitz/endpoints.py [hunk 1]
```diff
@@ -113,34 +122,47 @@ def get_current_events():
...
+    updated_ids = set()
 
-    events = current_app.event_manager.events
+    while True:
+        updated = current_app.updater.get_event_update()
+        if updated:
+            updated_ids.add(updated)
+            continue
+        if updated is False:
+            break
```
Comment (lunkwill42) [inline src/howitz/endpoints.py:134]: While `is False` may work here, you are in fact making an *identity* comparison with the built-in object `False`, not a value comparison.  The called function may actually return the bool object `False`, but you shouldn't rely on it when making boolean comparisons.   `if not updated` would be the idiomatic way.

This whole section would read better as 
```suggestion
        if not updated:
            break
        updated_ids.add(updated)
```

**Review by lunkwill42 — CHANGES_REQUESTED** [review]
My, how much simpler this gets when zinolib can handle the updates :)

File: src/howitz/endpoints.py (hunk 1 above)
Comment (lunkwill42) [reply src/howitz/endpoints.py:151]: @podliashanyk other than the code being just two lines of simple iterators, no.

Comment (lunkwill42) [reply src/howitz/endpoints.py:151]: I'd add that in 99% of cases you will ever need the `is` keyword in Python, it is when explicitly comparing a value to `None`, as in `x is None` or `x is not None`.

**Review by hmpf — COMMENTED** [review]
I'll check the functionality.

This needs a round of ruff or black. (My eyes hurt a bit.) We could do that for the entire codebase after this PR is in.

**Review by hmpf — APPROVED** [review]
I have done som emanual testing. Looks good!

File: requirements-frozen.txt [hunk 2]
```diff
@@ -48,5 +48,5 @@ werkzeug==2.3.8
     #   flask
     #   flask-login
     #   howitz (pyproject.toml)
-zinolib==1.0.1
+zinolib==1.0.4
```
Comment (lunkwill42) [inline requirements-frozen.txt:51]: Something is weird with this.  `pyproject.toml` says `zinolib>=1.1.0`, so why is zinolib `1.0.4` in the frozen requirements list?

**Review by lunkwill42 — CHANGES_REQUESTED** [review]
My nitpicks have been mostly fixed, but something seems off with the project requirements?  @hmpf?

File: requirements-frozen.txt (hunk 2 above)
Comment (hmpf) [reply requirements-frozen.txt:51]: Because we forgot to update it probably.

Comment (lunkwill42) [reply requirements-frozen.txt:51]: With `zinolib==1.0.4` I can't even get Howitz to run, with `1.1.0` this NTIE PR seems to work just fine for me :)

**Review by lunkwill42 — APPROVED** [review]
Works for me :)

**Pull request merged**